### ISSUE TITLE
lightningd.service config for encrypted hsm_secret

### ISSUE THREAD

**Issue by GitterDone221** [ElementsProject/lightning#6914]
I would like to set up a systemd lightning.service to ensure that upon system restart lightningd is automatically restarted.
Importantly, for my use case, I have an encrypted hsm_secret and would need to somehow incorporate the retrieval and entry of the hsm_secret password a part of the lightningd.service configuration. While I have found online examples of lightningd.service config, none address how to handle lightningd start-up with encrypted hsm_secret.

Could you please provide a sample lightnind.service config for use with encrypted hsm_secret.

**Comment by cdecker** [comment]
I don't think we can provide a canonical way of doing this, as the strategies to manage and secure passwords are very varied. The fundamental problem is that if we were to just add the password in a file, without requiring user interaction, then that would be no more secure that keeping the hsm_secret unencrypted.

If you can share a bit more about how you'd like to manage the password (1password, pass, yubikey, ...) I think we can discuss this in a more concrete way. I think the `ExecStartPre` could be used to initiate a decrypt that waits for the user to e.g. press the yubikey, unlocking the secret and store it in an envvar or similar.

**Comment by 1ma** [comment]
Systemd 251+ supports credential handling natively. It encrypts these credentials with the TPM and 4096 bytes of entropy stored at /var/lib/systemd/credential.secret.

With `systemd-creds setup` the credential.secret file is generated, and with `systemd-creds encrypt --pretty` we can encrypt plaintext secrets in a format that is suitable to embed in world-readable service files.

```shell
# systemd-creds setup
4096 byte credentials host key set up.

#  echo -n "mysupersecretpassword" | systemd-creds encrypt --name="HSM_SECRET_FILE_PASSWORD" --pretty - -
SetCredentialEncrypted=HSM_SECRET_FILE_PASSWORD: \
        Whxqht+dQJax1aZeCGLxmiAAAAABAAAADAAAABAAAAC9k5ZTe4As2yc+WG0AAAAAF0WTD \
        RoHqs6g0ELduBXx/MSErAdqdgdVXOFAVNFZ0/lRVRfbG5OuafVDIGDH64u2fLwpvdm/L5 \
        80PKHjBZ0UBak41jFJvX4oC8aF4ZiLY2QAspA2Ng==
```

I think we only need CLN to retrieve the password from a "well known" credential name such as HSM_SECRET_FILE_PASSWORD, if it is available.

https://wiki.archlinux.org/title/Systemd-creds
https://systemd.io/CREDENTIALS/
https://commandmasters.com/commands/systemd-creds-linux/

**Comment by cdecker** [comment]
That is indeed a very intriguing idea. I'm wondering though if that isn't equivalent to having the secret on disk, but read only to the CLN user. Is there a way to require a physical interaction to release a given secret? That'd be cool, and similar to yubikeys and other security tokens. Otherwise, being a user authorized to retrieve the secret is similar to the on-disk case I think (disk swapping attacks being the obvious standout where a tpm can help but in that case if encrypt the entire disk).

Passing the secret to the binary is also fraught with pitfalls. Environment variables for example may show up in the process's /proc tree, and depending on distro are or are not protected. I'd likely go the other way around: rather than push from secret provider to CLN, have CLN pull in the secret by specifying a command to retrieve it at startup. This way the secret only ever lives in the hsmd memory.

**Comment by 1ma** [comment]
I think this could be achieved. The systemd-creds method is not exactly equivalent to having the plaintext secret on disk with readonly perms to the CLN user because the on-disk file is encrypted at all times (with a random key, or the TPM2 or both). Unfortunately I'm not aware of any method to integrate with YubiKey or similar, but in my case I'm looking to automate the bootstrap of a systemd service with an encrypted hsm_secret anyway.

When a service is started with a credential, systemd decrypts it into another file but also makes sure that that new file only lives under `/run/credentials/...` with restrictive permissions, and mounted on an unswappable ramfs so that it never touches the disk.

Additionally systemd passes a `CREDENTIALS_DIRECTORY` env var to the service where all the secrets can be found by name.

This is how I think systemd-creds could be leveraged by CLN:

1. Create the encrypted password file:
```shell
$ systemd-ask-password -n | sudo systemd-creds encrypt --name=hsm-password - /opt/hsm.pwd
🔐 Password: ************************
Credential secret file '/var/lib/systemd/credential.secret' is not located on encrypted media, using anyway.
```
systemd-ask-password is just a convenient way to input the password interactively and feed it through stdin to the next process, -n just strips the final newline. The warning about credential.secret can be avoided by setting up full disk encryption when installing the OS, or by encrypting the secret with the TPM (if one is available in the machine):

```shell
$ systemd-creds has-tpm2
yes
+firmware
+driver
+system
+subsystem
+libraries

$ systemd-ask-password -n | sudo systemd-creds encrypt --with-key=tpm2 --name=hsm-password - /opt/hsm.pwd
🔐 Password: ************************
```

2. Add `LoadCredentialEncrypted=hsm-password:/opt/hsm.pwd` to the CLN service unit. For demonstration purposes here I use `systemd-run` to run throwaway services and peek inside. We can call `systemd-creds list` to show the secrets available to that service.

```shell
$ sudo systemd-run -P --wait -p User=cln -p LoadCredentialEncrypted=hsm-password:/opt/hsm.pwd systemd-creds list
Running as unit: run-u232.service
NAME         SECURE SIZE PATH                                          
hsm-password secure  24B /run/credentials/run-u232.service/hsm-password
Finished with result: success
Main processes terminated with: code=exited/status=0
Service runtime: 15ms
CPU time consumed: 13ms
```

We can look at the credential file directly and see that it lives in a ramfs filesystem and it's only readable by the process user:

```shell
$ sudo systemd-run -P --wait -p User=cln -p LoadCredentialEncrypted=hsm-password:/opt/hsm.pwd mount | grep "/credentials/run"
Running as unit: run-u234.service
ramfs on /run/credentials/run-u234.service type ramfs (ro,nosuid,nodev,noexec,relatime,mode=700)
Finished with result: success
...

$ sudo systemd-run -P --wait -p User=cln -p LoadCredentialEncrypted=hsm-password:/opt/hsm.pwd ls -l '$CREDENTIALS_DIRECTORY'
Running as unit: run-u233.service
total 16
-r-------- 1 cln root 24 Mar 24 17:51 hsm-password
Finished with result: success
...
```

3. CLN would then pull the secret from that file instead of pushing it with an env var or similar. The logic could be something like `On startup: if encrypted-hsm is on, the $CREDENTIALS_DIRECTORY env var is defined and $CREDENTIALS_DIRECTORY/hsm-password is a readable file, read it and use it as the hsm password instead of asking for it interactively`.

A more general way to do that could be adding an option `encrypted-hsm-password-file` that you could point to any file (in this example `$CREDENTIALS_DIRECTORY/hsm-password`). But it's easy to foresee that node runners would probably misuse this and store their HSM password in plaintext, possibly also on non-encrypted hard drives.

**Comment by cdecker** [comment]
Oh cool, I still believe that the TPM isn't adding anything on top of what an on-disk file that is root-ro isn't already solving, unless we force some user interaction, but we can fight over opinions all day.

I wonder if a symlink to the decrypted securely stored file from `systemd-creds` wouldn't be sufficient. It points nowhere until the secret gets decrypted, and then it just uses it unencrypted.

**Comment by 1ma** [comment]
Ok, I think I've arrived at a much better solution that doesn't need `systemd-creds`, doesn't store files on-disk, doesn't require changes in CLN and relies on user interaction to start lightngind (!!)

The idea is to use Systemd to create a UNIX socket and link it to the stdin of the lightningd.service, then the password can be written to the socket interactively.

/etc/systemd/system/lightningd.socket:
```
[Socket]
ListenFIFO=%t/lightningd.socket
Service=lightningd.service
```
/etc/systemd/system/lightningd.service is the same as the one [in the repo](https://github.com/ElementsProject/lightning/blob/master/contrib/init/lightningd.service), I just added these lines in the Service section:
```
[Service]
...

Sockets=lightningd.socket
StandardInput=socket
StandardOutput=journal
StandardError=journal
...
```

After running `systemctl start lightningd.service` the process will wait to read the password from stdin if the hsm_secret was encrypted. Then the password can be provided via the socket:

```shell
$ systemd-ask-password | tee /run/lightningd.socket > /dev/null
🔐 Password: ******
```

Alternatively a simple "echo" is enough, but care must be taken to not store the command in the shell history (by prepending at least 1 space before the command):

```shell
$  echo "secretpasswd" > /run/lightningd.socket
```

To check that everything works the output of lightningd can be monitored the whole time in another shell with `journalctl -f -u lightningd.service`.

@cdecker if you think it's worthwhile I can add a Systemd section to the documentation detailing all this.

cc @GitterDone221

**Comment by sangbida** [comment]
This issue is being closed as part of issue cleanup.

**Reason**: About deprecated feature

If you believe this issue should remain open, please comment and we can reopen it.
